### PR TITLE
br: add more retryable error for retryer

### LINKED ISSUE #42909: br: full backup failed 'failed to flush checkpoint: RequestError: send request failed'
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)
br full backup

### 3. What did you see instead (Required)
[2023/04/10 22:56:33.828 +00:00] [ERROR] [main.go:59] ["br failed"] [error="failed to flush checkpoint: RequestError: send request failed\ncaused by: Put \"http://minio-service:9000/longrun/full-0410/checkpoints/data/_OaWPJXhpUk5pisCUsZZ8PQ0nKShz735dqO44vMAltg%3D_4077345761290549635.cpt\": read tcp 10.233.93.42:36378->10.111.102.127:9000: read: connection reset by peer"] [errorVerbose="RequestError: send request failed\ncaused by: Put \"http://minio-service:9000/longrun/full-0410/checkpoints/data/_OaWPJXhpUk5pisCUsZZ8PQ0nKShz735dqO44vMAltg%3D_4077345761290549635.cpt\": read tcp 10.233.93.42:36378->10.111.102.127:9000: read: connection reset by peer\ngithub.com/pingcap/errors.AddStack\n\t/go/pkg/mod/github.com/pingcap/errors@v0.11.5-0.20221009092201-b66cddb77c32/errors.go:174\ngithub.com/pingcap/errors.Trace\n\t/go/pkg/mod/github.com/pingcap/errors@v0.11.5-0.20221009092201-b66cddb77c32/juju_adaptor.go:15\ngithub.com/pingcap/tidb/br/pkg/storage.(*S3Storage).WriteFile\n\t/home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/br/br/pkg/storage/s3.go:499\ngithub.com/pingcap/tidb/br/pkg/checkpoint.(*CheckpointRunner).doFlush\n\t/home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/br/br/pkg/checkpoint/checkpoint.go:520\ngithub.com/pingcap/tidb/br/pkg/checkpoint.(*CheckpointRunner).startCheckpointRunner.func1\n\t/home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/br/br/pkg/checkpoint/checkpoint.go:378\nruntime.goexit\n\t/usr/local/go/src/runtime/asm_amd64.s:1598\nfailed to flush checkpoint"] [stack="main.main\n\t/home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/br/br/cmd/br/main.go:59\nruntime.main\n\t/usr/local/go/src/runtime/proc.go:250"]
### 4. What is your TiDB version? (Required)

sh-5.1# /br --version
Release Version: v7.0.0
Git Commit Hash: fc166ffae8576ac3efb2f5abac848d5edbaf2657
Git Branch: heads/refs/tags/v7.0.0
Go Version: go1.20.2
UTC Build Time: 2023-03-20 06:56:18
Race Enabled: false

### PR DESCRIPTION
**Pull request by Leavrth**
### What problem does this PR solve?

Issue Number: close #42909 

Problem Summary:
some error is retryable, such as:
1. `io.EOF` for getting ts from PD
2. `read: connection reset` for connecting to s3
### What is changed and how it works?
modify the `shouldRetry` logic
### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: br/pkg/utils/backoff.go [hunk 1]
```diff
@@ -194,11 +194,11 @@ func (bo *pdReqBackoffer) NextBackoff(err error) time.Duration {
 	// bo.attempt--
 	e := errors.Cause(err)
 	switch e { // nolint:errorlint
-	case nil, context.Canceled, context.DeadlineExceeded, io.EOF, sql.ErrNoRows:
+	case nil, context.Canceled, context.DeadlineExceeded, sql.ErrNoRows:
```
Comment (YuJuncen) [inline br/pkg/utils/backoff.go:197]: I think we'd better to figure out why `io.EOF` isn't retryable from beginning.

**Review by YuJuncen — APPROVED** [review]

File: br/pkg/utils/backoff.go (hunk 1 above)
Comment (Leavrth) [reply br/pkg/utils/backoff.go:197]: I guess just move code from lightning.

**Review by 3pointer — COMMENTED** [review]
can we add a test case here?

**Review by 3pointer — APPROVED** [review]

**Pull request merged**